Add new gall guide. (#1249)
urbit/urbit.org: content/docs/userspace/gall-guide/4-lifecycle.md

+++
title = "4. Lifecycle"
weight = 20
template = "doc.html"
+++

> In the last lesson we look at a couple of useful things used as boilerplate in

Reviewer: ```suggestion
In the last lesson we looked at a couple of useful things used as boilerplate in
```

## Lifecycle

An agent's lifecycle starts when it's first installed. At this point, the
agent's `on-init` arm is called. This is the _only_ time `on-init` is ever
called - its purpose is just to initialize the agent. The `on-init` arm might be
very simple and just set an initial value for the state, or even do nothing at
all and return the agent core exactly as-is. It may also be more complicated,
and perform some [scries](/docs/arvo/concepts/scry) to obtain extra data or
check that another agent is also installed. It might send off some `card`s to
other agents or vanes to do things like load data in to the `%settings-store`
agent, bind an Eyre endpoint, or anything else. It all depends on the needs of
your particular application. If `on-init` fails for whatever reason, the agent
installation will fail and be aborted.

Once initialized, an agent will just go on doing its thing - processing events,
updating its state, producing effects, etc. At some point, you'll likely want to
push an update for your agent. Maybe it's a bug fix, maybe you want to add extra
features. Whatever the reason, you need to change the source code of your agent,

> so you commit a modified version of the file to Clay. When this happens, three things happen:

Reviewer: ```suggestion
so you commit a modified version of the file to Clay. When the commit completes, Gall updates the app as follows:
```

- The agent's `on-save` arm is called, which packs the agent's state in a `vase`
  and exports it.
- The new version of the agent is built and loaded into Gall.
- The previously exported `vase` is passed to the `on-load` arm of the newly
  built agent. The `on-load` arm will process it, convert it to the new version
  of the state if necessary, and load it back into the state of the agent.

A `vase` is just a cell of `[type-of-the-noun the-noun]` - it's used extensively

> for sending data between agents, vanes and ships. A vase is made with the

Reviewer: As noted in the types section, I don't think we send vases around between ships.

We'll look at the three arms described here in a little more detail, but first
we need to touch on the state itself.

## Versioned state type

In the previous lesson we introduced the idea of composing additional cores into
the subject of the agent core. Here we'll look at using such a core to define
the type of the agent's state. In principle, we could make it as simple as this:

```hoon
|%
+$  my-state-type  @ud
--
```

However, when you update your agent as described in the [Lifecycle](#lifecycle)
section, you may want to change the type of the state itself. This means
`on-load` might find old versions of the state in the `vase` it receives. Over
time, as you make further changes to the type of the state, this becomes quite
complicated.

For example, if you were creating an agent for a To-Do task management app, your
tasks might initially have a simple `?` to specify whether they're complete or
not. Something like:

```hoon
(map task=@t done=?)
```

At some point, you might want to add a third status to represent "in progress",
which might involve changing `done` to a tagged union like:

```hoon
(map title=@t status=?(%todo %done %work))
```

> The conventional way to keep this managable is to have _versioned states_. The

Reviewer: I think the explanation above doesn't cover the primary reason for adding a version tag to your state type: it makes it possible to distinguish between versions when that otherwise might not have been possible.

For example, say I update state from `counter=@ud` to `(map term @ud)`. If the old state is `0`, is that the empty map, or the `counter=0`? There's no way to know. (Even more obvious, versions where the state type doesn't change, but we do want special transition logic to occur, for example to recover from a now-fixed bug.)

resumed or reinstalled.

The state is packed in a vase with the
[zapgar](/docs/hoon/reference/rune/zap#-zapgar) (`!>`) rune, like `!>(state)`.

### `on-load`

This arm takes a `vase` and produces a `(quip card _this)`. Its purpose is to
import a state previously exported with [`on-save`](#on-save). Typically
you'd have used a [versioned state](#versioned-state-type) as described above,
so this arm would test which state version the imported data has, convert data
from an old version to the new version if necessary, and load it into the
`state` wing of the subject.

The vase would be unpacked with a
[zapgal](/docs/hoon/reference/rune/zap#-zapgal) (`!<`) rune, and then typically
you'd test its version with a [wuthep](/docs/hoon/reference/rune/wut#--wuthep)
(`?-`) expression.

## Example

Here's a new agent to demonstrate the concepts we've discussed here:

```hoon
/+  default-agent, dbug
|%
+$  versioned-state
  $%  state-0
  ==
+$  state-0  [%0 val=@ud]
+$  card  card:agent:gall
--
%-  agent:dbug
=|  state-0
=*  state  -
^-  agent:gall
|_  =bowl:gall
+*  this  .
    def   ~(. (default-agent this %.n) bowl)
++  on-init
  ^-  (quip card _this)

>   `this(val 42)

Reviewer: I continue to feel weird about this notation and prefer `[~ this(val 42)]` for clarity, but also concede that the pattern here is common enough. Just 2c

++  on-save
  ^-  vase
  !>(state)
++  on-load
  |=  old-state=vase
  ^-  (quip card _this)
  =/  old  !<(versioned-state old-state)
  ?-  -.old
    %0  `this(state old)
  ==

> ++  on-poke   on-poke:def

Reviewer: Now that these are getting actual implementations, consider separating the arms out like so for enhanced style & readability:

```suggestion
::
++  on-save
  ^-  vase
  !>(state)
::
++  on-load
  |=  old-state=vase
  ^-  (quip card _this)
  =/  old  !<(versioned-state old-state)
  ?-  -.old
    %0  `this(state old)
  ==
::
++  on-poke   on-poke:def
```

arm of a trap like `$(a b, c d, ...)`. It's the same concept here - we're saying
`this` (our agent core) with `val` replaced by `42`. Since `on-init` is only
called when the agent is first installed, we're just initializing the state.

Next we have `on-save`:

```hoon
++  on-save
  ^-  vase
  !>(state)
```

This exports our agent's state, and is called during upgrades, suspensions, etc.
We're having it pack the `state` value in a `vase`.

Finally, we have `on-load`:

```hoon
++  on-load
  |=  old-state=vase
  ^-  (quip card _this)
  =/  old  !<(versioned-state old-state)
  ?-  -.old
    %0  `this(state old)
  ==
```

It takes in the old state in a `vase`, then unpacks it to the `versioned-state`
type we defined earlier. We test its head for the version, and load it back into
the state of our agent if it matches. This test is a bit redundant at this stage
since we only have one state version, but you'll soon see the purpose of it.

You can save it as `/app/lifecycle.hoon` in the `%base` desk and `|commit %base`. Then, run `|rein %base [& %lifecycle]` to start it.

Let's try inspecting our state with `dbug`:

```
>   [%0 val=42]

> > :state-test +dbug

Reviewer 2: ```suggestion
> :lifecycle +dbug
```

>=
```

`dbug` can also dig into the state with the `%state` argument, printing the value of the specified face:

```
>   42

> > :state-test +dbug [%state %val]

Reviewer 2: ```suggestion
> :lifecycle +dbug [%state %val]
```

++  on-fail   on-fail:def
--
```

As soon as you `|commit` it, Gall will immediately export the existing state
with `on-save`, build the new version of the agent, then import the state back
in with `on-load`.

In the state definition core, you'll see we've added a new state version with a different structure:

```hoon
+$  versioned-state
  $%  state-0
      state-1
  ==
+$  state-0  [%0 val=@ud]
+$  state-1  [%1 val=[@ud @ud]]
+$  card  card:agent:gall
--
```

We've also changed the part that adds the state, so it uses the new version instead:

```hoon
=|  state-1
=*  state  -
```

In `on-init`, we've updated it to initialize the state with a value that fits the new type we've defined:

```hoon
++  on-init
  ^-  (quip card _this)
  `this(val [27 32])
```

`on-init` won't be called in this case, but if someone were to directly install
this new version of the agent, it would be, so we still need to update it.

`on-save` has been left unchanged, but `on-load` has been updated like so:

```hoon
++  on-load
  |=  old-state=vase
  ^-  (quip card _this)
  =/  old  !<(versioned-state old-state)
  ?-  -.old
    %1  `this(state old)
    %0  `this(state 1+[val.old val.old])
  ==
```

> We've updated the wuthep expression with a new case that handles our new state

Reviewer: Honestly a bit surprised at the amount of times you write out the rune's pronounceable name instead of just putting the rune itself. No strong feelings (though certainly here `?-` would be easier to link back to the code), but wondering what your motivation for that is, if any.

```suggestion
We've updated the `?-` expression with a new case that handles our new state
```

type, and for the old state type we've added a function that converts it to the
new type - in this case by duplicating `val` and changing the head-tag from `%0`
to `%1`. This is an extremely simple state type transition function - it would
likely be more complicated for an agent with real functionality.

> Let's now use `dbug` to check our state has successfully been updated to the new

Reviewer: ```suggestion
Let's now use `dbug` to confirm our state has successfully been updated to the new
```

type:

```
>   [%1 val=[42 42]]
> :lifecycle +dbug
>=
```

## Summary

- The app lifecycle rougly consists of initialization, state export, upgrade,
  state import and state version transition.
- This is managed by three arms: `on-init`, `on-save` and `on-load`.
- `on-init` initializes the agent and is called when it's first installed.
- `on-save` exports the agent's state and is usually called during upgrade or
  when an app is suspended.
- `on-load` imports an agent's state and is usually called during upgrade or
  when an app is unsuspended. It typically also handles converting data from old

>   state versions to new state versions.

Reviewer: Feel like the writing here could be a tad tighter.

```suggestion
- `on-save` exports the agent's state and is called during upgrade or
  when an app is suspended.
- `on-load` imports an agent's state and is called during upgrade or
  when an app is unsuspended. It also handles converting data from old
  state versions to new state versions.
```
